src/ngraph/runtime/gpu/cudnn_emitter.cpp

```diff
@@ -146,29 +184,30 @@ size_t runtime::gpu::CUDNNEmitter::build_reduce_forward(const runtime::gpu::GPUR
         *ctx->cudnn_handle, desc, input_desc, output_desc, &workspace_size));
     size_t workspace_idx = allocator.reserve_workspace(workspace_size);
 
+    void* alpha = get_data_by_type(data_type, 1.0);
```

Comment:
Why this instead of declaring on stack and using the captured variables address like before? Also, I don't see alpha and beta being freed/deleted anywhere.

Reply:
Since cudnnDataType_t is a constant expression in these use cases. Can we do this at compile time instead of runtime? It's a little bit more code, but it offloads the resolution to the compiler instead of doing it at runtime and avoids the need for heap allocations/deallocations. Something like this:

```C++
template <cudnnDataType_t data_type>
struct type_from_cudnn_datatype
{
        static_assert(not (data_type == CUDNN_DATA_HALF), "CUDNN_DATA_HALF is not supported by cuDNN");
        static_assert(not (data_type == CUDNN_DATA_INT8x4), "CUDNN_DATA_INT8x4 is not supported by cuDNN");
};

template<>
struct type_from_cudnn_datatype<CUDNN_DATA_FLOAT>
{
        static float type(double value) { return static_cast<float>(value); }
};

template<>
struct type_from_cudnn_datatype<CUDNN_DATA_DOUBLE>
{
        static double type(double value) { return static_cast<double>(value); }
};

template<>
struct type_from_cudnn_datatype<CUDNN_DATA_INT32>
{
        static int32_t type(double value) { return static_cast<int32_t>(value); }
};

template<>
struct type_from_cudnn_datatype<CUDNN_DATA_INT8>
{
        static int8_t type(double value) { return static_cast<int8_t>(value); }
};

template<cudnnDataType_t data_type>
auto get_data_by_type(double value) -> decltype(type_from_cudnn_datatype<data_type>::type(value)) {
        return type_from_cudnn_datatype<data_type>::type(value);
}

int main()
{

        float float_variable = get_data_by_type<CUDNN_DATA_FLOAT>(42.5);
        int int_variable = get_data_by_type<CUDNN_DATA_INT32>(42.5);

        auto half_variable = get_data_by_type<CUDNN_DATA_HALF>(42.5); // compile-time error
        auto int8x4_variable = get_data_by_type<CUDNN_DATA_INT8x4>(42.5); // compile-time error

        std::cout << float_variable << " " << int_variable << std::endl;
}
```

Reply:
My understanding is we don't know the datatype at compile time.

Reply:
Ah, now that I look at it again, I see you are right. I thought the types were being hard coded in a switch, within which we could know the type at compile time.

Reply:
I'm trying to add some code to manage allocations/deallocations.

Reply:
one way is to make the whole build function as template<datatype>, and invoke the function by using a switch for different datatype.

Reply:
Add host parameters manager.